c7n/resources/redshift.py

```diff
@@ -603,6 +605,84 @@ def process(self, clusters):
         return clusters
 
 
+@Redshift.action_registry.register('set-attributes')
+class RedshiftSetAttributes(BaseAction):
+    """
+    Action to modify Redshift clusters
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+                - name: redshift-modify-cluster
+                  resource: redshift
+                  filters:
+                    - type: value
+                      key: AllowVersionUpgrade
+                      value: false
+                  actions:
+                    - type: set-attributes
+                      attributes:
+                        AllowVersionUpgrade: true
+    """
+
+    schema = type_schema('set-attributes',
+                        attributes={"type": "object"},
+                        required=('attributes',))
+
+    permissions = ('redshift:ModifyCluster',)
+
+    shape = 'ModifyClusterMessage'
+
+    def validate(self):
+        attrs = dict(self.data.get('attributes'))
+        if attrs.get('ClusterIdentifier'):
+            raise PolicyValidationError('ClusterIdentifier field cannot be updated')
+        attrs["ClusterIdentifier"] = ""
+        return shape_validate(attrs, self.shape, 'redshift')
+
+    def process(self, clusters):
+        client = local_session(self.manager.session_factory).client(
+            self.manager.get_model().service)
+        for cluster in clusters:
+            self.process_cluster(client, cluster)
+
+    def process_cluster(self, client, cluster):
+        try:
+            config = dict(self.data.get('attributes'))
+            modify = {}
+            for k, v in config.items():
+                if k == 'ElasticIp' and v != jmespath.search('ElasticIpStatus.ElasticIp', cluster):
+                    modify[k] = v
+                elif (k == 'ClusterSecurityGroups' and v !=
+                jmespath.search('ClusterSecurityGroups[].ClusterSecurityGroupName', cluster)):
+                    modify[k] = v
+                elif (k == 'VpcSecurityGroupIds' and v !=
+                jmespath.search('VpcSecurityGroups[].VpcSecurityGroupId', cluster)):
+                    modify[k] = v
+                elif (k == 'HsmClientCertificateIdentifier' and
+                v != jmespath.search('HsmStatus.HsmClientCertificateIdentifier', cluster)):
+                    modify[k] = v
+                elif (k == 'HsmConfigurationIdentifier' and
+                v != jmespath.search('HsmStatus.HsmConfigurationIdentifier', cluster)):
+                    modify[k] = v
+                elif v != cluster.get('PendingModifiedValues').get(k, cluster.get(k)):
+                    modify[k] = v
+            if not modify:
+                return
+
+            modify['ClusterIdentifier'] = (cluster.get('PendingModifiedValues')
+                                          .get('ClusterIdentifier')
+                                          or cluster.get('ClusterIdentifier'))
+            client.modify_cluster(**modify)
+        except Exception as e:
```

Comment:
not found should be handled separately then the generic exception, also should only catch ClientError as the generic exception here